include/sys/queued_operation.h

```diff
@@ -0,0 +1,239 @@
+	 * operation is acceptable only if a non-negative value is
+	 * returned.
+	 *
+	 * If not provided queued_operation_submit() will assume
+	 * service-specific expectations are trivially satisfied, and
+	 * will reject the operation only if the notification
+	 * configuration is unacceptable.
+	 *
+	 * @param mgr the service that supports queued operations.
+	 *
+	 * @param op the operation being considered for suitability.
+	 *
+	 * @return the value to be returned from queued_operation_submit().
+	 */
+	int (*validate)(struct queued_operation_manager *mgr,
+			struct queued_operation *op);
+
+	/**
+	 * @brief Function to transform a generic notification
+	 * callback to its service-specific form.
+	 *
+	 * The implementation should cast cb to the proper signature
+	 * for the service, and invoke the cast pointer with the
+	 * appropriate arguments.
+	 *
+	 * @param mgr the service that supports queued operations.
+	 *
+	 * @param op the operation that has been completed.
+	 *
+	 * @param cb the generic callback to invoke.
+	 */
+	void (*callback)(struct queued_operation_manager *mgr,
+			 struct queued_operation *op,
+			 async_notify_generic_callback cb);
+
+	/**
+	 * @brief Function used to inform the manager of a new operation.
+	 *
+	 * This is called as a side effect of
+	 * queued_operation_submit() or queued_operation_finalize() to
+	 * tell the service that a new operation needs to be
+	 * processed, or that there are no operations left to process.
+	 * The function will not be invoked while an operation is in
+	 * progress.
+	 *
+	 * @param mgr the service that supports queued operations.
+	 *
+	 * @param op the operation that should be initiated.  A null
+	 * pointer is passed if there are no pending operations.
+	 */
+	void (*process)(struct queued_operation_manager *mgr,
+			struct queued_operation *op);
+};
+
+/**
+ * @brief State associated with a manager instance.
+ */
+struct queued_operation_manager {
+	/* Links the operation into the operation queue. */
+	sys_slist_t operations;
+
+	/* Pointer to the functions that support the manager. */
+	const struct queued_operation_functions *vtable;
+
+	/* Lock controlling access to other fields. */
+	struct k_spinlock lock;
+
+	/* The operation that is being processed. */
+	struct queued_operation *current;
+
+	/* A flag indicating that the last operation has been
+	 * completed and the callback notification is being
+	 * executed.
+	 */
+	bool finalizing;
+};
+
+#define QUEUED_OPERATION_MANAGER_INITIALIZER(_vtable) {	\
+		.vtable = _vtable,			\
+}
+
+/**
+ * @brief Submit an operation to be processed when the service is
+ * available.
+ *
+ * During the call to this function the service process function will
+ * have been invoked at least once, either providing another operation
+ * or indicating that no operations are pending.
+ *
+ * @param mgr a generic pointer to the service instance
+ * @param op a generic pointer to an operation to be performed
```

Comment:
I think that it is worth mentioning that `notify` field in `struct queued_operation` must be initialized prior calling submit using async_notify_init api.